openedx/core/djangoapps/content_libraries/api.py

```diff
@@ -1051,3 +1089,285 @@ def revert_changes(library_key):
+
+
+# Import from Courseware
+# ======================
+
+
+class BaseEdxImportClient(abc.ABC):
+    """
+    Base class for all courseware import clients.
+
+    Import clients are wrappers tailored to implement the steps used in the
+    import APIs and can leverage different backends.  It is not aimed towards
+    being a generic API client for Open edX.
+    """
+
+    EXPORTABLE_BLOCK_TYPES = {
+        "drag-and-drop-v2",
+        "problem",
+        "html",
+        "video",
+    }
+
+    def __init__(self, library_key=None, library=None):
+        """
+        Initialize an import client for a library.
+
+        The method accepts either a library object or a key to a library object.
+        """
+        if bool(library_key) == bool(library):
+            raise ValueError('Provide at least one of `library_key` or '
+                             '`library`, but not both.')
+        if library is None:
+            library = ContentLibrary.objects.get_by_key(library_key)
+        self.library = library
+
+    @abc.abstractmethod
+    def get_block_data(self, block_key):
+        """
+        Get the block's OLX and static files, if any.
+        """
+
+    @abc.abstractmethod
+    def get_export_keys(self, course_key):
+        """
+        Get all exportable block keys of a given course.
+        """
+
+    @abc.abstractmethod
+    def get_block_static_data(self, asset_file):
+        """
+        Get the contents of an asset_file..
+        """
+
+    def import_block(self, modulestore_key):
+        """
+        Import a single modulestore block.
+        """
+
+        # Get or create the block in the library.
+
+        block_data = self.get_block_data(modulestore_key)
+
+        try:
+            library_block = create_library_block(
+                self.library.library_key,
+                modulestore_key.block_type,
+                modulestore_key.block_id,
+            )
+            blockstore_key = library_block.usage_key
+        except LibraryBlockAlreadyExists:
+            blockstore_key = LibraryUsageLocatorV2(
+                lib_key=self.library.library_key,
+                block_type=modulestore_key.block_type,
+                usage_id=modulestore_key.block_id,
+            )
+            get_library_block(blockstore_key)
+
+        # Handle static files.
+
+        for filename, static_file in block_data.get('static_files', {}).items():
+            files = [
+                f.path for f in
+                get_library_block_static_asset_files(blockstore_key)
+            ]
+            if filename in files:
+                # Files already added, move on.
+                continue
+            file_content = self.get_block_static_data(static_file)
+            add_library_block_static_asset_file(
+                blockstore_key, filename, file_content)
+
+        # Import OLX and publish.
+
+        set_library_block_olx(blockstore_key, block_data['olx'])
+        publish_changes(blockstore_key.lib_key)
```

Comment:
@jvdm, can you please pull this in?  https://github.com/open-craft/edx-platform/pull/379/commits/ee4c4c469a8600bd2a23812ae3a829c145cac847

For performance reasons, we can only publish at the end of the process.